LSPS4/README.md

```diff
@@ -0,0 +1,1923 @@
+> using C libraries, which may also trigger the same undefined
+> behavior on signed integer overflow, thus unsigned integers are
+> specified, which, in C, are specifically modulo 2 to the number
+> of bits.
+> Arithmetic overflow is problematic and may lead to unexpectedly
+> low `opening_fee` charges for high `payment_size_msat` and high
+> `opening_fee_params.proportional`, thus must be specifically
+> detected and cause an abort in this flow rather than continue
+> and cause problems for the LSP.
+> Most modern microprocessors have dedicated instructions to detect
+> overflow, and most modern compilers can recognize the above code
+> and optimize those down to these dedicated instructions at high
+> optimization settings; in particular, the division in the
+> overflow-detecting multiplication routine is optimized away
+> and replaced with a simple overflow-flag check.
+
+### 6. Buying Inbound Liquidity
+
+The client selects a set of PPPs it wants to have forwarded
+to it.
+
+The total amounts of the PPPs would now be the `payment_size`.
+The client then computes the `opening_fee`.
+
+A portion of the PPPs will be used to pay the `opening_fee`,
+while the total of the PPPs (i.e. the `payment_size_msat`)
+determines how large the `opening_fee` would be.
+The larger the total amount of the PPPs selected, the higher
+the `opening_fee` due to the proportional part.
+
+The client MUST check if the `opening_fee` is strictly less
+than `payment_size - number_of_ppps * client_htlc_minimum_msat`,
+where `number_of_ppps` is the number of PPPs in the set and
+`client_htlc_minimum_msat` is the corresponding setting in the
+`opening_fee_params`.
+
+The client MUST check that `client_htlc_minimum_msat` is strictly
+less than or equal to the smallest PPP amount in the selected
+PPP set.
+
+Once the client has selected the set of PPPs it will use to
+pay for the channel open, the client calls `lsps4.open_with_ppps`
+call, which accepts the parameters:
+
+```JSON
+{
+  "ppp_ids": ["id#abcdefghijklmnop"],
+  "opening_fee_params": {
+    "min_fee_msat": "546000",
+    "proportional": 1000,
+    "client_htlc_minimum_msat": 100,
+    "valid_until": "2023-04-19T16:55:21.295Z",
+    "min_lifetime": 1008,
+    "max_client_to_self_delay": 1008,
+    "promise": "abcdefghijklmnopqrstuvwxyz+/"
+  }
+}
+```
+
+All parameters are required.
+
+`ppp_ids` is a set of PPP string identifiers that the client wants
+to use to open the channel.
+It MUST NOT be empty.
+
+`opening_fee_params` is an object selected from the result of
+the `lsps4.get_opening_fee` call, copied verbatim.
+
+The entire `lsps4.open_with_ppps` call MUST be atomic, including
+the checks below.
+
+The LSP, atomically:
+
+* MUST check that there is currently no pending LSPS4-bought
+  channel currently being opened (i.e. has not completed up to
+  `channel_ready` from both parties).
+  * This is done by checking for the existence of a
+    "PPP Opening Set", described later.
```

Comment:
Is this strictly necessary?
The main business model of an LSP is to open channels for fees.
As long as this request is not using the same payment parts as a previous request, this should be fine?

Reply:
LSP can more easily limit its risk by always handling just one channel open at a time, as well as simplify implementation.  If client cannot rely on an LSP being able to handle multiple channel opens, then it is better if we say that here.

There is a reason why we have a `operation_pending` error code for this.